### PR TITLE
always start dependency task when requested, terminating existing / in progress ones to prevent race condition

### LINKED ISSUE #203776: Rerun VSCode build tasks often only restarts one task
I am seeing this since months but more frequently lately. I often rerun build tasks via this keybinding:

```json
{
    "key": "cmd+shift+alt+b",
    "command": "workbench.action.tasks.runTask",
    "args": "Restart VS Code - Build"
}
```

In 1 out of 2 cases only one task restarts:

image

Is there a place I can check for logs to diagnose further or can we add some logging?

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #203776

Tested in a custom build based on this branch

### CODE REVIEW

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 1]
```diff
@@ -570,6 +575,7 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 		const count = lastInstance?.count ?? { count: 0 };
 		count.count++;
 		const activeTask = { task, promise, count };
+		this._log('setting active task ' + activeTask.task._label);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:578]:
```suggestion
```

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 2]
```diff
@@ -1051,7 +1056,10 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 				this._terminalService.focusInstance(terminal);
 			}
 		}
-		this._activeTasks[task.getMapKey()].terminal = terminal;
+		if (this._activeTasks[task.getMapKey()]) {
+			// it's possible this has been removed already
+			this._activeTasks[task.getMapKey()].terminal = terminal;
+		}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1062]: Put in a variable?

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 3]
```diff
@@ -1051,7 +1056,10 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 				this._terminalService.focusInstance(terminal);
 			}
 		}
-		this._activeTasks[task.getMapKey()].terminal = terminal;
+		if (this._activeTasks[task.getMapKey()]) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1059]: Can the types for these string dictionaries be fixed to catch this at compile time? Something like this?

```
private _activeTasks: IStringDictionary<IActiveTerminalData | undefined>;
private _busyTasks: IStringDictionary<Task | undefined>;
private _terminals: IStringDictionary<ITerminalData | undefined>;
```

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1059]: I think we always want to remove the task's map key vs keeping it in the map with a value of `undefined`?

Comment (Tyriar) [reply src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts]: You may attempt to keep it in sync, but it's possible to get undefined so the compiler should know that and warn us if we may be causing an NPE

**Review by Tyriar — APPROVED** [review]

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts (hunk 3 above)
Comment (Tyriar) [reply src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts]: Approved, but this is a good debt task which could catch some errors

**Pull request merged**